storage/pipeline/receive.go

```diff
@@ -0,0 +1,287 @@
+package sealing
+
+import (
+	"bytes"
+	"context"
+	"net/url"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-datastore"
+	"github.com/multiformats/go-multihash"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/crypto"
+	"github.com/filecoin-project/go-state-types/proof"
+	"github.com/filecoin-project/go-statemachine"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/storage/sealer/storiface"
+)
+
+func (m *Sealing) Receive(ctx context.Context, meta api.RemoteSectorMeta) error {
+	m.inputLk.Lock()
+	defer m.inputLk.Unlock()
+
+	si, err := m.checkSectorMeta(ctx, meta)
+	if err != nil {
+		return err
+	}
+
+	exists, err := m.sectors.Has(uint64(meta.Sector.Number))
+	if err != nil {
+		return xerrors.Errorf("checking if sector exists: %w", err)
+	}
+	if exists {
+		return xerrors.Errorf("sector %d state already exists", meta.Sector.Number)
+	}
+
+	err = m.sectors.Send(uint64(meta.Sector.Number), SectorReceive{
+		State: si,
+	})
+	if err != nil {
+		return xerrors.Errorf("receiving sector: %w", err)
+	}
+
+	return nil
+}
+
+func (m *Sealing) checkSectorMeta(ctx context.Context, meta api.RemoteSectorMeta) (SectorInfo, error) {
```

Comment:
So checking this is actually more than a little tricky:
* Because of datastore prunning the receiving node may not have this message stored
* Even if it did, it's not necessarily a SubmitCommit[Batch] message
  * It can be a multisig tx
  * (after fvm m2) It can be some other message to a proxy actor

And in all cases if SaaS is able to send the commit message miners behalf, the miner probably trusts the service provider quite a lot, at least enough to not put malicious cids in sector metadata.

We can check the cid prefix easily though.

Reply:
Ok makes sense!